clients/src/main/java/org/apache/kafka/common/network/SslTransportLayer.java

```diff
@@ -240,40 +240,47 @@ protected boolean flush(ByteBuffer buf) throws IOException {
     */
     @Override
     public void handshake() throws IOException {
-        int read = 0;
+        // Reset state to support renegotiation. This can be removed if renegotiation support is removed.
+        if (state == State.READY)
+            state = State.HANDSHAKE;
+
         // Read any available bytes before attempting any writes to ensure that handshake failures
         // reported by the peer are processed even if writes fail (since peer closes connection
         // if handshake fails)
+        int read = 0;
         if (key.isReadable())
             read = readFromSocketChannel();
 
         try {
             doHandshake();
         } catch (SSLException e) {
-            handleSslException(e, true);
+            handshakeFailure(e, true);
         } catch (IOException e) {
+            maybeThrowSslAuthenticationException();
+
             // this exception could be due to a write. If there is data available to unwrap,
             // process the data so that any SSLExceptions are reported
             if (handshakeStatus == HandshakeStatus.NEED_UNWRAP && netReadBuffer.position() > 0) {
                 try {
                     handshakeUnwrap(false);
                 } catch (SSLException e1) {
-                    handleSslException(e1, false);
+                    handshakeFailure(e1, false);
                 }
             }
             // If we get here, this is not a handshake failure, throw the original IOException
             throw e;
         }
 
-        // Read failed, but there were no other exceptions while performing handshake, so throw EOFException
-        if (read == -1)
+        // Read from socket failed, so throw any pending handshake exception or EOF exception.
+        if (read == -1) {
+            maybeThrowSslAuthenticationException();
             throw new EOFException("EOF during handshake, handshake status is " + handshakeStatus);
+        }
     }
 
     private void doHandshake() throws IOException {
         boolean read = key.isReadable();
         boolean write = key.isWritable();
-        state = State.HANDSHAKE;
```

Comment:
Maybe we should call `maybeThrowSslAuthenticationException` instead of doing the check manually a few lines below.